bigflow/workflow.py

```diff
@@ -23,24 +30,52 @@ def daily_start_time(start_time: dt.datetime) -> dt.datetime:
 
 
 class Workflow(object):
-    def __init__(self,
-                 workflow_id,
-                 definition,
-                 schedule_interval=DEFAULT_SCHEDULE_INTERVAL,
-                 start_time_factory: typing.Callable[[dt.datetime], dt.datetime] = daily_start_time,
-                 ):
+
+    def __init__(
+        self,
+        workflow_id,
+        definition,
+        schedule_interval=DEFAULT_SCHEDULE_INTERVAL,
+        start_time_factory: typing.Callable[[dt.datetime], dt.datetime] = daily_start_time,
+        log_config: typing.Optional[log.LogConfigDict] = None,
+    ):
         self.definition = self._parse_definition(definition)
         self.schedule_interval = schedule_interval
         self.workflow_id = workflow_id
         self.start_time_factory = start_time_factory
 
+        self._log_initialized = False
+        self.log_config = log_config
+
+    def init_logging(self):
+        """Initialize python logging based on the configuration attached to the workflow. """
+
+        if self._log_initialized:
+            return
+        self._log_initialized = True
+
+        if not self.log_config:
+            print("Log configuration is not provided: skip")
+            return
+        log.init_logging(
+            self.log_config,
+            workflow_id=self.workflow_id,
+        )
+
+    def _check_logging_initialized(self):
+        if self.log_config and not self._log_initialized:
+            import warnings
+            warnings.warn(UserWarning("Log configuration is attached to the Workflow, but method `.init_logging() is not called"))
```

Comment:
So maybe if config provided init_logging() should be called in __init__

Reply:
IMO it is not a best place for it.

Reply:
Why

Reply:
Because:
- It might be hard to test code, where __init__ contains some side effects (modification of env, logging etc)
- bigflow command tries to walk across all python modules, which triggers creation of all workflows in project - so all Workflow.__init__ will be called multiple times "by-design"
- actually putting this into __ init __ will mean "initialization in the Workflow", which was discussed in other comment-thread.